### ISSUE TITLE
Support OOPIF

### ISSUE THREAD

**Issue by aslushnikov** [puppeteer/puppeteer#2548]
[Out-of-process iframes](https://www.chromium.org/developers/design-documents/oop-iframes) are here now and have to be supported.

For the following HTML served from `localhost`, pptr fails to report the frame:

```html
<h1>Hello world</h1>
<script>
  const frame = document.createElement('iframe')
  frame.setAttribute('name', 'bob')
  frame.setAttribute('src', 'https://google.com')
  document.body.appendChild(frame)
</script>
```

Original example: https://github.com/achingbrain/puppeteer-frames
Reported here: https://github.com/GoogleChrome/puppeteer/issues/1140#issuecomment-389054750

**Comment by achingbrain** [comment]
It doesn't even need any JavaScript on the page, plain HTML also fails in the same way:

```html
<!DOCTYPE html>
<html>
  <head>
    <title>Where is my iframe?</title>
  </head>
  <body>
    <h1>Hello world</h1>
    <iframe name="bob" src="https://google.com"></iframe>
  </body>
</html>
```

**Comment by aslushnikov** [comment]
While we're figuring this out, a workaround is to run with `--disable-features=site-per-process` flag:

```js
const browser = await puppeteer.launch({                                                                                                                                                                                                                                                            
  args: ['--disable-features=site-per-process'],                                                                                                                                                                                                                                                    
  headless: false                                                                                                                                                                                                                                                                                   
}) ;
```

**Comment by tomholub** [comment]
Unfortunately `--disable-features=site-per-process` does not help for extension frames on standard domains. Still missing. Tested with bundle from #2506

**Comment by aslushnikov** [comment]
@tomholub indeed, thanks for pointing out. It turned out extension isolation is hard-coded in [`chrome/browser/extensions/chrome_content_browser_client_extensions_part.cc`](https://cs.chromium.org/chromium/src/chrome/browser/extensions/chrome_content_browser_client_extensions_part.cc?type=cs&g=0&l=366-385).

**Comment by aslushnikov** [comment]
For the record: puppeteer v1.5.0 disables out-of-process iframes by default (see #2661). This, however, doesn't apply to extension iframes - these still operate out-of-process.

**Comment by Swordyjohn** [comment]
Is there a time frame for OOPIF support?

**Comment by aslushnikov** [comment]
@Swordyjohn no good eta; is this pressing you in any way?

**Comment by Swordyjohn** [comment]
@aslushnikov we are attempting to use Puppeteer for scraping a site with strong content security in a business setting (I'm an RPA in background screening) and we have had more success than other methods but we cannot access reCaptcha src info when served in order to export it to a click farm due to the OOPIF so we will probably try different automation frameworks.

**Comment by tomholub** [comment]
It's pressing us too - cannot test extension frames inside non extension
tabs.

**Comment by Banou26** [comment]
Any updates on this ? I really need OOPIF.
Any way i can help on this ?

**Comment by defr0std** [comment]
Not sure if it solves a problem with extensions, but here is what seems to be the proper way to work with out-of-process iframes.
https://bugs.chromium.org/p/chromium/issues/detail?id=924937#c13

**Comment by benbuzbee** [comment]
@aslushnikov could be willing to take a stab at this but how long this issue has been neglected makes me nervous that there may be some hidden dangers.

**What would you say are the major concerns?**

The change seems something like
- use target attached/detatched/destroyed events with type=iframe as a hint to the frame tree and update the frame manager accordingly
https://cs.chromium.org/chromium/src/content/browser/devtools/protocol/target_auto_attacher.cc?type=cs&sq=package:chromium&g=0&l=173

- The frame needs its own CDPSession if created in this way instead of inheriting the parents so that it can use a separate sessionid, seems relatively straight-forward. Seems to be some finagling to do here so each instance ignores messages not meant for it

- Commands that use the devtools Page domain that aren't already on the Frame class may need to be changed to send to all frame's sessions
  _This one may be sketch. Some commands will be fine, like captureScreenshot which can exist only being dispatched to the root frame, but others wont. It looks like Page.addScriptToEvaluateOnNewDocument doesn't filter to OOPIF so Page::exposeFunction() won't work for them.  That said, this limitation seems like something that could be slowly addressed over time in the future and not having it but having most other things seems strictly better._

**Comment by benkauffman** [comment]
+1

**Comment by bbracha-evinced** [comment]
+1

**Comment by Banou26** [comment]
Don't do that guys... It's useless, just use the reactions instead...
All it does is notify people watching this thread for nothing, no value added

**Comment by JoseBarrios** [comment]
I ran into the same issue. I found a workaround/hack that might help folks here:

```Js

  /**                                                                                                    
   * Waits until the iframe is attached and then returns it to the caller                        
   *                                                                                                     
   * @access public
   * @param {object} page - The Puppeteer page API object 
   * @param {string} nameOrId - The name or id of the target iframe                                                                                                                                      
   * @returns {object} The Puppeteer iframe element                                                            
   */                                                                                                    
  async function iframeAttached(page, nameOrId) {                                                                  
    return new Promise(async resolve => {
      const pollingInterval = 1000;                                                                
      const poll = setInterval(async function waitForIFrameToLoad() {                                                                                                       
        const iFrame = page.frames().find(frame => frame.name() === nameOrId);                         
        if (iFrame) {                                                                                    
          clearInterval(poll);                                                                                  
          resolve(iFrame);                                                                               
        }                                                                                                                                                                         
      }, pollingInterval);                                                                                          
    });                                                                                                  
  } 
```

Now, you can get your iframes like this:

```Js
const iframe = await iframeAttached(page, 'YOUR_IFRAME_NAME_OR_ID');
// You can now use the iframe
iframe.doSomething(...);
```

**Comment by clouedoc** [comment]
Should this issue be closed by #7556 ?
